AMDGPU: Make v2f64 -> v2f16 conversion Legal only when unsafe fast math is set

changpeng opened the pull request:
On targets that support v_cvt_pk_f16_f32 instruction, if we make v2f64 -> v2f16 Legal, we will generate the following sequence of instructions:
   v_cvt_f32_f64_e32 v1, s[6:7]
   v_cvt_f32_f64_e32 v2, s[4:5]
   v_cvt_pk_f16_f32 v1, v2, v1
While it may be when unsafe fast math is set, it possibly returns imprecise results due to double rounding. This patch makes v2f64 -> v2f16 conversion Legal only when unsafe fast math is set.

Fixes: SWDEV-523856

changpeng committed:
AMDGPU: Make v2f64 -> v2f16 conversion Legal only for unsafe fast math
  On targets that support v_cvt_pk_f16_f32 instruction, if we make v2f64 -> v2f16
Legal, we will generate the following sequence of instructons:
   v_cvt_f32_f64_e32 v1, s[6:7]
   v_cvt_f32_f64_e32 v2, s[4:5]
   v_cvt_pk_f16_f32 v1, v2, v1
While it is okay when unsafe fast math flag is set, it may return imprecise results
due to double rounding. This patch makes v2f64 -> v2f16 conversion Legal only when
unsafe fast math is set.

Fixes: SWDEV-523856

changpeng commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -915,7 +915,7 @@ SITargetLowering::SITargetLowering(const TargetMachine &TM,
 
-  if (Subtarget->hasCvtPkF16F32Inst()) {
+  if (Subtarget->hasCvtPkF16F32Inst() && TM.Options.UnsafeFPMath) {
     setOperationAction(ISD::FP_ROUND, MVT::v2f16, Legal);
```
> It disables it not just from v2f64, but also from v2f32, right? How is that lowered then?
We have a pattern to generate v_cvt_pk_f16_f32:
https://github.com/llvm/llvm-project/blob/9b63a92ca723293dfe8570d1b2881ce949f1f6cc/llvm/lib/Target/AMDGPU/VOP3Instructions.td#L1194

Even though we also have a pattern for v2f64->v2f16, but we will do splitting the vector and then Custom lowering of f64 ->f16 first.
https://github.com/llvm/llvm-project/blob/9b63a92ca723293dfe8570d1b2881ce949f1f6cc/llvm/lib/Target/AMDGPU/VOP3Instructions.td#L1196

rampitec replied:
But the pattern shall not work for operation which is not legal.

changpeng replied:
> But the pattern shall not work for operation which is not legal.

my observation is that, if we do not Expand/Custom/Promote, the pattern will be picked up, no matter you specify Legal or not. So it doesn't matter whether you specify Legal or not, pattern for v2f32->v2f16 wil always be picked up.

This is similar to f32 -> f16, even though we did not specify Legal, v_cvt_f16_f32 will still be selected.

changpeng replied:
Default is Legal?

changpeng replied:
I think Legal is default: 
 https://llvm.org/docs/WritingAnLLVMBackend.html#legal

The reason we need to explicitly specify Legal sometimes is because we need to specify different Actions based on different hardware features.

changpeng replied:
I am still learning. I think even though we specify Custom lowering but actually doing nothing, like:
if (SrcVT. != MVT::f64)
    return Op;  // **f32 is legal for fp_round**
This is essentially "Legal", which means either there is a native instruction, or a pattern to match to native instructions.  The compiler will throw an error if there is nothing to select.

arsenm reviewed (changes requested):
Should not further spread use of the global options. FP casts have fast math flags, make use of those. Note this needs to be in custom lowering, legality cannot be use context dependent

arsenm commented on llvm/test/CodeGen/AMDGPU/fptrunc.v2f16.fpmath.ll:
```diff
@@ -0,0 +1,199 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx950 -global-isel=0 -enable-unsafe-fp-math < %s | FileCheck -check-prefixes=GFX950,GFX950-UNSAFE %s
+; RUN: llc -mtriple=amdgcn -mcpu=gfx950 -global-isel=1 -enable-unsafe-fp-math < %s | FileCheck -check-prefixes=GFX950,GFX950-UNSAFE %s
+
+define amdgpu_ps <2 x half> @v_test_cvt_v2f32_v2f16(<2 x float> %src) {
```
Don't need the shared conventions. Also test the scalar case and wider vector s

changpeng replied:
> Don't need the shader conventions. Also test the scalar case and wider vector s

Removed the shade conventions, and added tests for other types.

changpeng committed:
AMDGPU: Make v2f64 -> v2f16 conversion Legal only for unsafe fast math

  Custom lower FP_ROUND for v2f16. Also add test cases for scalar and wider vectors

changpeng commented:
> Should not further spread use of the global options. FP casts have fast math flags, make use of those. Note this needs to be in custom lowering, legality cannot be use context dependent

I understand your proposal to avoid use of global options. But here we need to be consistent. Options.UnsafeFPMath has been used in all other places to expand/lower f64->f16. We should not use a different approach only here.

changpeng commented:
> > Should not further spread use of the global options. FP casts have fast math flags, make use of those. Note this needs to be in custom lowering, legality cannot be use context dependent
> 
> I understand your proposal to avoid use of global options. But here we need to be consistent. Options.UnsafeFPMath has been used in all other places to expand/lower f64->f16. We should not use a different approach only here.

Or do you prefer a combination of fast math flag and global option, like this?
const SDNodeFlags Flags = Op->getFlags();
  bool AllowInaccurateFP_ROUND =
      Flags.hasApproximateFuncs() || DAG.getTarget().Options.UnsafeFPMath;

arsenm commented:
> I understand your proposal to avoid use of global options. But here we need to be consistent. Options.UnsafeFPMath has been used in all other places to expand/lower f64->f16. We should not use a different approach only here.

This is because traditionally the FP casts did not have fast math flags. They do now, and we should migrate to only using them.

arsenm commented:
> Or do you prefer a combination of fast math flag and global option, like this? const SDNodeFlags Flags = Op->getFlags(); bool AllowInaccurateFP_ROUND = Flags.hasApproximateFuncs() || DAG.getTarget().Options.UnsafeFPMath;

You can do that, but it still needs to be a contextual use transformation, not modify the legality rules

changpeng commented:
> > Or do you prefer a combination of fast math flag and global option, like this? const SDNodeFlags Flags = Op->getFlags(); bool AllowInaccurateFP_ROUND = Flags.hasApproximateFuncs() || DAG.getTarget().Options.UnsafeFPMath;
> 
> You can do that, but it still needs to be a contextual use transformation, not modify the legality rules

I do not fully understand what did you mean "modify the legality rules", and why we should not modify.

Here is the example in custom lowering for FP_TO_FP16: 

  if (getTargetMachine().Options.UnsafeFPMath) {
    // There is a generic expand for FP_TO_FP16 with unsafe fast math.
    return **SDValue();**
  }

Do you mean this is to change from Custom to Expand, and is not allowed?   Thanks.

changpeng commented:
> > I understand your proposal to avoid use of global options. But here we need to be consistent. Options.UnsafeFPMath has been used in all other places to expand/lower f64->f16. We should not use a different approach only here.
> 
> This is because traditionally the FP casts did not have fast math flags. They do now, and we should migrate to only using them.
Do you mean the instruction fptrunc can has "fast" flag now? 
%res = fptrunc **fast** <2 x double> %src to <2 x half>

How to check the instruction has the "fast" flag?  
Flags.hasApproximateFuncs() is not working when we have the "fast"   Thanks

changpeng commented:
> > I understand your proposal to avoid use of global options. But here we need to be consistent. Options.UnsafeFPMath has been used in all other places to expand/lower f64->f16. We should not use a different approach only here.
> 
> This is because traditionally the FP casts did not have fast math flags. They do now, and we should migrate to only using them.

On SelectDAG path, FP_ROUND still does not have fast math flags. I still think adding fast math flags for FP_ROUND is beyond the work of this patch. (We also need to propagate the flags when we do transformation like splitting the vector).

changpeng committed:
AMDGPU: Make v2f64 -> v2f16 conversion Legal only when unsafe fast math is set

  Custom lowering v2f64 -> v2f16.

changpeng commented:
Custom lowering for v264 -> v2f16 on both SelectionDAG and GlobalISel paths now. Fast math flag for fp_round is completely not working and needs additional work beyond this patch.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -1053,10 +1053,10 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
-  else
+  if (ST.hasCvtPkF16F32Inst()) {
+    FPTruncActions.legalFor({{S32, S64}, {S16, S32}, {V2S16, V2S32}})
+                  .customFor({V2S16, V2S64});
```
It's suspect that we need custom handling for the v2 case, but not the regular scalar

changpeng replied:
> It's suspect that we need custom handling for the v2 case, but not the regular scalar

Previously only v2f16 was made Legal in order to generate v_cvt_pk_f16_f32, I think.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -2741,6 +2743,29 @@ bool AMDGPULegalizerInfo::legalizeMinNumMaxNum(LegalizerHelper &Helper,
+bool AMDGPULegalizerInfo::legalizeFPTrunc(LegalizerHelper &Helper,
+                                          MachineInstr &MI,
+                                          MachineRegisterInfo &MRI) const {
+  // TODO: We should only use fast math flag. But the global option is
+  // still used here to be consistent, especially when the fast math flag is
+  // not working for FP_ROUND on the SelectDAG path at this moment.
+  MachineFunction &MF = Helper.MIRBuilder.getMF();
+  bool AllowInaccurateFPTRUNC = MI.getFlag(MachineInstr::FmAfn) ||
+                                MF.getTarget().Options.UnsafeFPMath;
+
+  if (AllowInaccurateFPTRUNC) {
+    // Use the tablegen pattern to select native instructions.
+    return true;
+  }
+
+  Register DstReg = MI.getOperand(0).getReg();
+  LLT DstTy = MRI.getType(DstReg);
+
+  // Scalarize the vector and fall through to lower f64 -> f16.
+  return Helper.fewerElementsVector(MI, 0, DstTy.getElementType()) ==
+         LegalizerHelper::Legalized;
```
This seems like it should be a combine and not a modification of legalization rules

changpeng replied:
> This seems like it should be a combine and not a modification of legalization rules

I am looking at instcombine to generate v_cvt_pk_f16_f32 instructions.  Should we implement in AMDGPU because it is target dependent? Thanks.

arsenm commented on llvm/test/CodeGen/AMDGPU/fptrunc.v2f16.fpmath.ll:
```diff
@@ -0,0 +1,1748 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Test belongs in the existing fptrunc.f16.ll test

changpeng replied:
> Test belongs in the existing fptrunc.f16.ll test

The existing fptrunc.f16.ll only checks when unsafe fpmath is set.  This patch actually fixes the codegen when unsafe fpmath is not set for v2f16 only. Maybe we should not include scalar and wider in this file?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:
```diff
@@ -2741,6 +2743,29 @@ bool AMDGPULegalizerInfo::legalizeMinNumMaxNum(LegalizerHelper &Helper,
+    // Use the tablegen pattern to select native instructions.
+    return true;
+  }
+
```
This handling logic doesn't make sense. The decision to scalarize or not has nothing to do with the FP properties? The tablegen pattern also handles the build_vector case, which is the MIR I would expect to select from. Can we just drop the direct vector handling, and always scalarize? 

The tablegen patterns should also be guarded with the fast checks if we're going to have them.

changpeng replied:
> This handling logic doesn't make sense. The decision to scalarize or not has nothing to do with the FP properties? The tablegen pattern also handles the build_vector case, which is the MIR I would expect to select from. Can we just drop the direct vector handling, and always scalarize?
> 
> The tablegen patterns should also be guarded with the fast checks if we're going to have them.

The logic to generate instructions based on the pattern, but this is only correct when the unsafe fpmath is set.
Yes, we may always scalarize. Should we do instcombine to generate v_cvt_pk_f16_f32?

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -6893,6 +6893,19 @@ SDValue SITargetLowering::lowerFP_ROUND(SDValue Op, SelectionDAG &DAG) const {
 
   EVT DstVT = Op.getValueType();
+
+  if (DstVT == MVT::v2f16) {
```
Same comments as the globalisel path. The vectoriness shouldn't matter, we don't need to directly consume the operation

changpeng replied:
> Same comments as the globalisel path. The vectoriness shouldn't matter, we don't need to directly consume the operation

Okay, we can do as https://github.com/llvm/llvm-project/pull/135659, which essentially scalarize v2f16.  WE may do instcombine to geenrate the vector cvt when the target support. Thanks.
